### PR TITLE
add dense matmul benchmark

### PR DESCRIPTION
**Pull request by jorendumoulin**
This adds a benchmark that measures peak performance of the gemm accelerator measuring from the time it is launched until it is finished.

It is configured to run the benchmark on every push to main, and uploads the output report to github artifacts, sneek peak available here:   https://github.com/KULeuven-MICAS/snax-mlir/actions/runs/11012799975

### CODE REVIEW

File: benchmarks/dense_matmul/genbenchmark.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+from xdsl.builder import ImplicitBuilder
+from xdsl.dialects import arith, builtin, func, linalg
+from xdsl.dialects.builtin import i8, i32
+from xdsl.ir import Block, Region
+from xdsl.printer import Printer
+
+from util.snax_benchmark import SNAXBenchmark
+
+
+def create_tiled_matrix_multiply(k, m, n):
```
Comment (JosseVanDelm) [inline benchmarks/dense_matmul/genbenchmark.py:15]: This is not a tiled_matrix_multiply anymore :)

File: benchmarks/dense_matmul/genbenchmark.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    region = Region(b)
+
+    function = func.FuncOp.from_region("streamer_matmul", input_types, [], region)
+
+    module = builtin.ModuleOp([function])
+
+    return module
+
+
+def write_module_to_file(module, file):
```
Comment (JosseVanDelm) [inline benchmarks/dense_matmul/genbenchmark.py:60]: I want to move this to the tools folder at some point

File: benchmarks/dense_matmul/gendata.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    variables = {
+        "A": A,
+        "B": B_new_layout,
+        "C_golden": C_golden,
+        "C": C,
+    }
+
+    create_header("data.h", sizes, variables)
+    create_data("data.c", variables)
```
Comment (JosseVanDelm) [inline benchmarks/dense_matmul/gendata.py:50]: We can probably move this to a more common folder :)

File: benchmarks/dense_matmul/main.c [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      {
+        int32_t error = memrefC.aligned_data[i] - C_golden[i];
+        // printf("%d) %d -> %d\n", i, (int32_t)memrefC.aligned_data[i],
+        //        (int32_t)C_golden[i]);
+        if (error != 0)
+          nerr += 1;
+      }
+    }
+
+    // insert mcycle to show fault in trace
```
Comment (JosseVanDelm) [inline benchmarks/dense_matmul/main.c:79]: What do you mean by this?

**Review by JosseVanDelm — APPROVED** [review]
Yeah let's do this!
We can probably refactor a bunch of things out of here though.
Furthermore, can you refactor everything that says tiled_matmul to not say it?
This is not a tiled matrix multiplication after all ;)

File: benchmarks/dense_matmul/main.c (hunk 4 above)
Comment (jorendumoulin) [reply benchmarks/dense_matmul/main.c:79]: just a quick way to check if an operation succeeds by looking at the trace. if there are 3 mcycles, it was successful. if there are 4, an error has occured

**Pull request merged**